release/scripts/PowerShell/FhirServerRelease/Private/Grant-ClientAppDelegatedPermissions.ps1

```diff
@@ -0,0 +1,108 @@
+function Grant-ClientAppDelegatedPermissions {
+    <#
+    .SYNOPSIS
+    Grants delegated permissions to a client app, so that users of the app are 
+    not required to consent to the app calling the FHIR apli app on their behalf.
+    .PARAMETER AppId
+    The client application app ID.
+    .PARAMETER TenantAdminCredential
+    Credentials for a tenant admin user
+    .PARAMETER ResourceApplicationId
+    Application Id for the resource for which we need access
+    #>
+    param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AppId,
+
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNull()]
+        [pscredential]$TenantAdminCredential,
+
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNull()]
+        [string]$ResourceApplicationId
+    )
+
+    Set-StrictMode -Version Latest
+
+    Write-Host "Granting delegated permissions for app ID $AppId"
+
+    # Get token to talk to graph api
+    $tenantId = (Get-AzureADCurrentSessionInfo).TenantId.ToString()
+
+    $adTokenUrl = "https://login.microsoftonline.com/$tenantId/oauth2/token"
+    $resource = "https://graph.microsoft.com/"
+
+    $body = @{
+        grant_type = "password"
+        username   = $TenantAdminCredential.GetNetworkCredential().UserName
+        password   = $TenantAdminCredential.GetNetworkCredential().Password
+        resource   = $resource
+        client_id  = "1950a258-227b-4e31-a9cf-717495945fc2" # Microsoft Azure PowerShell
+    }
+
+    try {
+        $response = Invoke-RestMethod -Method 'Post' -Uri $adTokenUrl -ContentType "application/x-www-form-urlencoded" -Body $body -ErrorVariable error
+    }
+    catch {
+        Write-Warning "Failed to get auth token to talk to graph api."
+        Write-Warning "Error message: $error"
+
+        throw
+    }
+
+    $windowsAadId = "00000002-0000-0000-c000-000000000000"  #ResourceId for Windows Azure Active Directory
+    $windowsAadServicePrincipal = Get-AzureAdServicePrincipal -Filter "appId eq '$windowsAadId'"
+    $windowsAadObjectId = $windowsAadServicePrincipal.ObjectId
+
+    $resourceApiServicePrincipal = Get-AzureAdServicePrincipal -Filter "appId eq '$ResourceApplicationId'"
+    $resourceApiObjectId = $resourceApiServicePrincipal.ObjectId
+
+    $clientServicePrincipal = Get-AzureAdServicePrincipal -Filter "appId eq '$AppId'"
+    $clientObjectId = $clientServicePrincipal.ObjectId
+
+    $header = @{
+        'Authorization' = 'Bearer ' + $response.access_token
+        'Content-Type' = 'application/json'
+    }
+
+    $permissionGrantUrl = "https://graph.microsoft.com/v1.0/oauth2PermissionGrants"
+
+    $bodyForReadPermisions = @{
+        clientId = $clientObjectId
+        consentType = "AllPrincipals"
+        resourceId = $windowsAadObjectId
+        scope = "User.Read"
+    }
+
+    # TODO: Handle scenario where application might already have these permissions.
```

Comment:
What would happen currently in this situation? Would the step fail, or just not be needed?

Reply:
The api call returns a 400 :(. It shouldn't have an impact on our pipeline since we clear everything and setup a fresh test env every run. I have also added code that checks for existing permissions in case we run into an error.